server/src/main/java/org/elasticsearch/cluster/coordination/Coordinator.java

```diff
@@ -873,7 +873,15 @@ void becomeCandidate(String method) {
             joinAccumulator = joinHelper.new CandidateJoinAccumulator();
 
             peerFinder.activate(coordinationState.get().getLastAcceptedState().nodes());
-            clusterFormationFailureHelper.start();
+
+            // Only start the formation helper if NOT in shutdown.
+            // The shutdown process will disconnect the node from the cluster, and start
```

Comment:
Not sure I follow this comment.
I had an assumption that this error happens when both elected master and this node is shut down simultaneously.
Is it not the case?

Reply:
I'm not really clear what the contract is for shutdown. `becomeCandidate` is called from many places, and this block is `mode != Mode.CANDIDATE`. Before the node starts shutdown, it is not master, but is part of a cluster. The master node is fine and still running during this time -- it later removes the shut down node from the cluster.

This PR is related to ES-8412 (sorry I forgot to link it), which came out of an investigation (ES-6977) showing that shutdown begins and >10 seconds later we get a log message (`last failed join attempt was 18.9m ago`) from the ClusterFormationFailureHelper. No node-join request is ever sent, even though ClusterFormationFailureHelper gets started again (otherwise it wouldn't be logging msgs). My assumption is that shutdown eventually disconnects the node and causes the node to try to reconnect with the cluster.

```
stopping ...

Native controller process has stopped - no new native processes can be started

[3] consecutive checks of the master node [{es-es-index-7f4f764fcc-9mzns}{6t7LDGi4QDeEFJe__BtDoQ}{NJyqTYZSTz2AuKh_Q49lnQ}{es-es-index-7f4f764fcc-9mzns}{10.2.145.23}{10.2.145.23:9300}{Iimr}{8.15.0}{7000099-8505000}] were unsuccessful ([3] rejected, [0] timed out), restarting discovery; more details may be available in the master node logs [last unsuccessful check: node closed {es-es-index-7f4f764fcc-fkwqs}{2K0YJBgqTqKohccqANHoWA}{boU4zO8mRwqX649Wzk1P_Q}{es-es-index-7f4f764fcc-fkwqs}{10.2.146.58}{10.2.146.58:9300}{Iimr}{8.15.0}{7000099-8505000}{k8s_node_name=ip-10-104-28-134.ec2.internal, transform.config_version=10.0.0, ml.config_version=12.0.0, stateless.memory=68719476736, xpack.installed=true}]

retrying master election after [10] failed attempts; election attempts are currently scheduled up to [1100ms] apart

last failed join attempt was 18.9m ago, failed to join {es-es-index-7f4f764fcc-9mzns}{6t7LDGi4QDeEFJe__BtDoQ}{NJyqTYZSTz2AuKh_Q49lnQ}{es-es-index-7f4f764fcc-9mzns}{10.2.145.23}{10.2.145.23:9300}{Iimr}{8.15.0}{7000099-8505000}{stateless.memory=68719476736, transform.config_version=10.0.0, ml.config_version=12.0.0, k8s_node_name=ip-10-104-29-211.ec2.internal, xpack.installed=true} with JoinRequest{sourceNode={es-es-index-7f4f764fcc-fkwqs}{2K0YJBgqTqKohccqANHoWA}{boU4zO8mRwqX649Wzk1P_Q}{es-es-index-7f4f764fcc-fkwqs}{10.2.146.58}{10.2.146.58:9300}{Iimr}{8.15.0}{7000099-8505000}{k8s_node_name=ip-10-104-28-134.ec2.internal, transform.config_version=10.0.0, ml.config_version=12.0.0, stateless.memory=68719476736, xpack.installed=true}, compatibilityVersions=CompatibilityVersions[transportVersion=8643000, systemIndexMappingsVersion={.secrets-inference=MappingsVersion[version=1, hash=-1434574148], .fleet-agents-7=MappingsVersion[version=1, hash=-1302168228], .fleet-servers-7=MappingsVersion[version=1, hash=-916922632], .ml-config=MappingsVersion[version=1, hash=1455005892], .fleet-policies-leader-7=MappingsVersion[version=1, hash=-1108172796], .security-tokens-7=MappingsVersion[version=1, hash=576296021], .geoip_databases=MappingsVersion[version=1, hash=-305757839], .security-profile-8=MappingsVersion[version=2, hash=-909540896], .search-app-1=MappingsVersion[version=1, hash=-501711141], .fleet-artifacts-7=MappingsVersion[version=1, hash=-1593703898], .query-rules-1=MappingsVersion[version=1, hash=1647955624], .transform-internal-007=MappingsVersion[version=1, hash=1144737897], .fleet-enrollment-api-keys-7=MappingsVersion[version=1, hash=-840564854], .fleet-actions-7=MappingsVersion[version=1, hash=961753072], .tasks=MappingsVersion[version=0, hash=-945584329], .ml-meta=MappingsVersion[version=1, hash=976416370], .fleet-secrets-7=MappingsVersion[version=1, hash=-745394230], .security-7=MappingsVersion[version=1, hash=-1061511639], .connector-secrets-1=MappingsVersion[version=1, hash=-745394230], .logstash=MappingsVersion[version=1, hash=-1058806351], .ml-inference-000005=MappingsVersion[version=3, hash=919553140], .inference=MappingsVersion[version=1, hash=645017346], .async-search=MappingsVersion[version=0, hash=-1403744380], .fleet-policies-7=MappingsVersion[version=1, hash=762674907], .synonyms-2=MappingsVersion[version=1, hash=-888080772]}], features=[features_supported, license-trial-independent-version, health.dsl.info, data_stream.rollover.lazy, health.extended_repository_indicator, rrf_retriever_supported, desired_node.version_deprecated, index_size.supported, data_stream.lifecycle.global_retention, usage.data_tiers.precalculate_stats, standard_retriever_supported, data_stream.auto_sharding, retrievers_supported, knn_retriever_supported, stats.include_disk_thresholds], minimumTerm=1120, optionalJoin=Optional[Join[votingNode={es-es-index-7f4f764fcc-fkwqs}{2K0YJBgqTqKohccqANHoWA}{boU4zO8mRwqX649Wzk1P_Q}{es-es-index-7f4f764fcc-fkwqs}{10.2.146.58}{10.2.146.58:9300}{Iimr}{8.15.0}{7000099-8505000}{k8s_node_name=ip-10-104-28-134.ec2.internal, transform.config_version=10.0.0, ml.config_version=12.0.0, stateless.memory=68719476736, xpack.installed=true}, masterCandidateNode={es-es-index-7f4f764fcc-9mzns}{6t7LDGi4QDeEFJe__BtDoQ}{NJyqTYZSTz2AuKh_Q49lnQ}{es-es-index-7f4f764fcc-9mzns}{10.2.145.23}{10.2.145.23:9300}{Iimr}{8.15.0}{7000099-8505000}{k8s_node_name=ip-10-104-29-211.ec2.internal, transform.config_version=10.0.0, ml.config_version=12.0.0, stateless.memory=68719476736, xpack.installed=true}, term=1120, lastAcceptedTerm=1119, lastAcceptedVersion=1056459]]}

master not discovered or elected yet, an election requires a node with id [6t7LDGi4QDeEFJe__BtDoQ], have discovered possible quorum [{es-es-index-7f4f764fcc-fkwqs}{2K0YJBgqTqKohccqANHoWA}{boU4zO8mRwqX649Wzk1P_Q}{es-es-index-7f4f764fcc-fkwqs}{10.2.146.58}{10.2.146.58:9300}{Iimr}{8.15.0}{7000099-8505000}]; discovery will continue using [10.2.128.228:9300, 10.2.130.97:9300, 10.2.142.21:9300, 10.2.143.218:9300, 10.2.145.23:9300, 10.2.147.176:9300, 10.2.158.45:9300, 10.2.159.108:9300, 10.2.159.97:9300, 10.2.160.49:9300, 10.2.188.3:9300, 10.2.191.152:9300] from hosts providers and [{es-es-index-7f4f764fcc-ssl84}{LXDzj-XkSlaW9usInFiIDw}{UcqUjZzeSdG1lKTgFodwTg}{es-es-index-7f4f764fcc-ssl84}{10.2.133.158}{10.2.133.158:9300}{Iimr}{8.15.0}{7000099-8505000}, {es-es-index-7f4f764fcc-9mzns}{6t7LDGi4QDeEFJe__BtDoQ}{NJyqTYZSTz2AuKh_Q49lnQ}{es-es-index-7f4f764fcc-9mzns}{10.2.145.23}{10.2.145.23:9300}{Iimr}{8.15.0}{7000099-8505000}, {es-es-index-74c8796c46-zbbc9}{JU_8WPAQTjuVS1o5UMdZCw}{3j6tRnQLTwCX1ZCFNR5qOA}{es-es-index-74c8796c46-zbbc9}{10.2.135.219}{10.2.135.219:9300}{Iimr}{8.15.0}{7000099-8505000}, {es-es-index-7f4f764fcc-b9mnn}{Qb8SkDM5StOP2ZcAtHy1IQ}{gfRn7bLiSiKD6dLexeokLQ}{es-es-index-7f4f764fcc-b9mnn}{10.2.184.138}{10.2.184.138:9300}{Iimr}{8.15.0}{7000099-8505000}, {es-es-index-7f4f764fcc-fkwqs}{2K0YJBgqTqKohccqANHoWA}{boU4zO8mRwqX649Wzk1P_Q}{es-es-index-7f4f764fcc-fkwqs}{10.2.146.58}{10.2.146.58:9300}{Iimr}{8.15.0}{7000099-8505000}, {es-es-index-7f4f764fcc-qd42f}{dXDvWwAhRTmvipArNXXL0g}{_PuIXQSoSyChB5sRckvdfA}{es-es-index-7f4f764fcc-qd42f}{10.2.160.49}{10.2.160.49:9300}{Iimr}{8.15.0}{7000099-8505000}, {es-es-index-7f4f764fcc-xkpwz}{KTCuZG7JTpCWNaptw3w-9g}{9cwFpdDmTOqiGlZkqtUZ-g}{es-es-index-7f4f764fcc-xkpwz}{10.2.182.196}{10.2.182.196:9300}{Iimr}{8.15.0}{7000099-8505000}, {es-es-index-7f4f764fcc-rjr2b}{Mi_R46PESo2BRdAFAU26gw}{IoYLjCJ6RQeab4ezDFCPwQ}{es-es-index-7f4f764fcc-rjr2b}{10.2.159.157}{10.2.159.157:9300}{Iimr}{8.15.0}{7000099-8505000}, {es-es-index-7f4f764fcc-nq99j}{CL6FJIVJR4qzSp0lccAvFQ}{26vDd6NzTHqLRSptitXxCA}{es-es-index-7f4f764fcc-nq99j}{10.2.159.107}{10.2.159.107:9300}{Iimr}{8.15.0}{7000099-8505000}] from last-known cluster state; node term 1120, last-accepted version 1057716 in term 1120; for troubleshooting guidance, see https://www.elastic.co/guide/en/elasticsearch/reference/current/discovery-troubleshooting.html
failed to connect to {es-es-index-74c8796c46-2hznx}{K9-wjQ7hQBe9St5-0luZeA}{rWJ3W2cSTNy4jy2cgnOTsA}{es-es-index-74c8796c46-2hznx}{10.2.144.175}{10.2.144.175:9300}{Iimr}{8.15.0}{7000099-8505000}{xpack.installed=true, transform.config_version=10.0.0, ml.config_version=12.0.0, k8s_node_name=ip-10-104-29-240.ec2.internal, stateless.memory=34359738368} (tried [1] times)

stopped

....
```